docs/design_notes/lift.md

```diff
@@ -0,0 +1,355 @@
+The function output only changes when an input argument changes.
+Therefore, a function is guaranteed to behave deterministically.
+
+But pure functions offer another big advantage to JAX: specifically, they enable functional transformations.
+For example `jax.vmap(f)` will vectorize a function `f`.
+Because `f` cannot have side effects the vectorized/parallel version of `f` is well-defined. To see why we need this restriction, consider what happens if `f` would increment a counter or draw a random number.
+Would `f` draw the same or a different random number for each item in the vector?
+Would each item in the batch have its own counter or is the counter shared among the items?
+And in what order is the counter incremented if `f` is computed in parallel?
+The answer to all these questions is "it depends".
+The behavior is ambiguous and the functional constraint elegantly avoids this problem.
+
+Flax introduces a safe way to have limited randomness and stateful variables in a JAX-compatible form.
+The reason why the state in Flax is not problematic is because it is local: inside a Flax `Module` there are variables and PRNG sequences,
+but on the outside there are only JAX Arrays and PRNG keys.
+
+For most use cases, Flax is used to define models in a stateful way.
+Because a `Module` behaves like a pure function externally, we can fully utilize JAX with all of its transformations.
+There are, however, cases when we want to have the best of both worlds by using transformations and `Module` together.
+This design note explains how we extend JAX's functional transformation to work on `Module`s that have internal state and randomness.
+
+
+## Functionalization
+
+Before we jump into the details let's consider a simple example where we would like to use `vmap` inside a `Module`.
+
+First, we define a simple MLP without any transformations:
+
+```python
+import jax
+from jax import random, numpy as jnp
+from flax import linen as nn
+
+class MLP(nn.Module):
+  @nn.compact
+  def __call__(self, xs):
+    h = nn.Dense(4, name='hidden')(xs)
+    h = nn.relu(h)
+    return nn.Dense(1, name='out')(h)
+```
+
+Now what if we want to have separate MLP parameters for each item in `xs`?
+If this were "vanilla JAX" we could imagine writing something like `jax.vmap(apply_mlp)(mlp_params, xs)`.
+But doing something like this in Linen will actually fail:
+
+```python
+class NaiveVmapMLP(nn.Module):
+  @nn.compact
+  def __call__(self, xs):
+    mlp = MLP()
+    return jax.vmap(lambda mlp, x: mlp(x))(mlp, xs)  # fails
+```
+
+JAX will raise an error when `vmap` is used on `mlp` because it's not a JAX array or a simple container of arrays.
+We can not really blame JAX for refusing to perform this under-specified job.
+After all, it's not even clear what should happen here.
+The parameters inside the MLP are not even initialized yet and we will need a separate PRNG key for each group of parameters.
```

Comment:
I think it may be confusing how the PRNG getting into the picture here. I think that it is doing so indirectly, because when this model will later be initialized it will be passed a PRNG, and we would like to specify that we want to pass a different key to each element of the input vector.

Reply:
Added some more clarificaiton here